Introduce EventsExecutor implementation

bmartin427 opened the pull request:
Closes #1389

Commit:
Introduce EventsExecutor implementation (#1389)

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

bmartin427 commented:
Draft form of this feature to drive discussion only.  I've done some minimal work on this relative to the original code to get things in roughly the right shape; however the functional test currently seems to explode, and a bunch of the linters are unhappy.  Even once this nominally works, there's a fair bit more work to be done on it, starting with the removal of the boost dependency.

Commit:
Fix explosion for reference count updates without GIL

The previous version worked on 22.04+Iron, but fails on 24.04+Jazzy or Rolling.  It
seems like the behavior of std::bind() may have changed when asked to bind a py::object
to a callback taking a py::handle.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Fix ament linter complaints

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Switch to non-boost asio library

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Use internal _rclpy C++ types instead of jumping through Python hoops

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Reformat with clang-format, then uncrustify again

clang-format fixes things that uncrustify tends to leave alone, but then uncrustify seems slightly
unhappy with that result.  Also reflow comments at 100 columns.

This uses the .clang-format file from the ament-clang-format package, with the exception of
SortIncludes being set to false, because I didn't like what it tried to do with includes without
that override.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Respect init signal handling options

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Reconcile signal handling differences with SingleThreadedExecutor

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
test_executor.py: Add coverage for EventsExecutor

Tests that currently work are enabled, and those that don't are annotated what needs to be done before they can be enabled

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Make spin_once() only return after a user-visible callback

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Add get_nodes() method

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Add context management support

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

jmachowinski commented on rclpy/src/rclpy/events_executor/timers_manager.cpp:
```diff
@@ -0,0 +1,335 @@
+        std::string("Failed to set timer reset callback: ") + rcl_get_error_string().str);
+    }
+    timers_[timer] = ready_callback;
+    if (timers_.size() == WARN_TIMERS_COUNT) {
```
```suggestion
    if (timers_.size() >= WARN_TIMERS_COUNT) {
```

bmartin427 replied:
So with this change, you'd get the warning re-printed with each additional timer beyond 8, rather than just once, which I was trying to avoid.  Granted, nothing stops you from creating seven timers, and then repeatedly creating and destroying the 8th, and getting the warning repeatedly anyway.

jmachowinski replied:
I would consider more warning a better behavior. The rational being, that nodes tend to be spammy with their output, and you could easily miss the warning. Also the management should get worse with every additional added timer, therefore I would say its worth printing an additional warning.

jmachowinski commented on rclpy/src/rclpy/events_executor/timers_manager.cpp:
```diff
@@ -0,0 +1,335 @@
+      if (next_call_ns <= 0) {
+        // This just notifies RCL that we're considering the timer triggered, for the purposes of
+        // updating the next trigger time.
+        const auto ret = rcl_timer_call(timer_cb_pair.first);
```
This happens directly before the the execution of the timer in the c++ version. I wonder if doing it here introduces some unwanted behavior...

bmartin427 replied:
None that I've encountered yet?  It just seemed important to get started on the *next* interval as quickly as possible to keep the timer rate from drifting due to the latency in dispatching the timer.  Are there any specific issues you suspect with this?

jmachowinski replied:
I remember some test failing around this detail in my implementations, sadly I don't remember the details. I guess the behavior around cancel is broken now.
E.g.
A callback taking some time is running.
Timer becomes ready, is added to the ready queue
Some other callback cancels the timer
Timer is executed anyway

Is there a check if the timer was canceled directly before the execution, that I missed ?

The interval should not drift, if this call is issued later. The reason for this is, that the next trigger time is computed as : 
```
while(nextTime < now)
{
   nextTimer = nextTime + period;
}

``` 

I think there is another bug lurking here. If a timer takes longer than its period, your queue will fill up. You need to block the timer from recomputing the next execution time, until it was executed.

bmartin427 replied:
Ah gotcha, yeah it won't drift but it will potentially miss whole intervals, but perhaps that was intended per your concern about queuing multiple calls.

As far as canceling, I'm not sure even `SingleThreadedExecutor` is safe from races on that front:
* I think a cancel between the `is_timer_ready()` check in `_wait_for_ready_callbacks()` and the `call_timer_with_info()` call in `_take_timer()` will cause an uncaught `RCLError` exception to be thrown:
```
  File "/home/brad/rclpy_ws/install/rclpy/lib/python3.12/site-packages/rclpy/executors.py", line 380, in _take_timer
    info = tmr.handle.call_timer_with_info()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rclpy._rclpy_pybind11.RCLError: failed to call timer: timer is canceled, at ./src/rcl/timer.c:268
```
* A cancel any time after `call_timer_with_info()` will not prevent the callback already in motion from eventually being delivered

I agree though that spacing out the `rcl_timer_call()` from the callback dispatch does make this race more likely to hit though.  I'll go ahead and rearrange that per the other concerns above.

bmartin427 replied:
I think this should be resolved now.  Both the `rcl_timer_call()` and the user callback now occur from the same `asio::post()` event.

jmachowinski commented on rclpy/src/rclpy/events_executor/timers_manager.cpp:
```diff
@@ -0,0 +1,335 @@
+  }
+  it->second->RemoveTimer(timer);
+  if (it->second->empty()) {
+    clock_managers_.erase(it);
```
hm, isn't there a race here, that the asio callback still happens into the now deleted clock_manager ?

bmartin427 replied:
I think the timer's own callback is safe, because it calls a user callback directly without capturing the clock manager itself.  However, it may be possible for `HandleJump()` and `UpdateTimers()` to get called with a deleted `this` pointer, if you remove the last timer from a clock immediately after a callback was posted, and before the callback is dispatched.  I'll make a note to look at this closer later.

bmartin427 replied:
I think this is resolved now.

jmachowinski commented on rclpy/src/rclpy/events_executor/events_executor.cpp:
```diff
@@ -0,0 +1,959 @@
+  // objects you're working with.  We'll try to figure out what kind of stuff is hiding behind the
+  // abstraction by having the Waitable add itself to a wait set, then take stock of what all ended
+  // up there.  We'll also have to hope that no Waitable implementations ever change their component
+  // entities over their lifetimes.
```
Hm, the rclcpp layer explicitly forces the waitables, to implement events handling. I don't know much about the rclpy waitable internals, but the approach taken here, would break the rclcpp waitables, so we should check this... A good candidate to have a look at should be the action interfaces.

bmartin427 replied:
There are test cases in `test_events_executor.py` specifically for actions, so I know that rclpy `Waitable` type works, unless you think there's something specific missing from that test coverage.

Can you clarify what it is you're looking for here?  Are you wanting me to modify the `Waitable` interface to better support callbacks directly?  That's probably a bit more ambitious and invasive than I was hoping to be, but I can look into it if you think it's necessary.

jmachowinski replied:
The expire mechanism of the action server is strange in the way, that the timer callback does not expire the goals. Instead this happens if the timer is expired / ready in the waitable mechanism. We had problems in that area in the rclcpp implementation. Therefore I would check if this still works.

bmartin427 replied:
test_events_executor.py has coverage for this, unless I'm misunderstanding what you mean.
https://github.com/bmartin427/rclpy/blob/events_executor/rclpy/test/test_events_executor.py#L643-L650

Commit:
Remove mutex protection on nodes_ member access

It was being used only inconsistently, and on reflection it wasn't adding any protection beyond what
the GIL already provides.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Fix deadlock during shutdown()

Needed to release the GIL while blocking on another lock.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
A little further on using C++ _rclpy types instead of Python interface

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Fix issue with iterating through incomplete Tasks

Never bool-test a py::object::attr() return value without an explicit py::cast<bool>.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Add support for coroutines to timer handling

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Fix test_not_lose_callback() test to destroy entities properly

EventsExecutor was exploding because the test was leaving destroyed entities in the node by using an
incorrect API to destroy them.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Correct test that wasn't running at all, and remove EventsExecutor from it

* Test methods need to be prefixed with 'test_' in order to function.  This had been entirely dead
  code, and when I enabled it EventsExecutor deadlocked.

* On reflection, it seems like a deadlock is exactly what should be expected from what this test is
  doing.  Remove EventsExecutor from the test and document the problem.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

bmartin427 committed:
Warn on every timer added over the threshold, not just the first

Co-authored-by: Janosch Machowinski <jmachowinski@users.noreply.github.com>
Signed-off-by: Brad Martin <52003535+bmartin427@users.noreply.github.com>

Commit:
Keep rcl_timer_call() tightly coupled with user callback dispatch

This both prevents an issue where user callbacks could potentially queue up if they didn't dispatch
fast enough, and ensures that a timer that has passed its expiration without being dispatched yet
can still be canceled without the user callback being delivered later.

This commit also makes use of the new rcl API for querying the absolute timer expiration time,
instead of the relative number of nanoseconds remaining until it expires.  This should both make
things more accurate, and potentially reduce overhead as we don't have to re-query the current time
for each outstanding timer.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Protect against deferred method calls happening against a deleted ClockManager

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Add support for new TimerInfo data passed to timer handlers

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Simplify spin_once() implementation

This both reduces duplicate code now, and simplifies the asio interface used which would need
replacing.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Fix stale Future done callbacks with spin_until_future_complete()

This method can't be allowed to leave its Future done callback outstanding in case the method is
returning for a reason other than the Future being done.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Replace asio::io_context with a new EventsQueue object

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Use existing rclpy signal handling instead of asio

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Replace asio timers with a dedicated timer wait thread

This is dumb on its own, but it helps me move towards eliminating asio.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Correct busy-looping in async callback handling

This isn't ideal because there are some ways async callbacks could become unblocked which wouldn't
get noticed right away (if at all); however this seems to match the behavior of
SingleThreadedExecutor.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Add EventsExecutor to new test_executor test from merge

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

bmartin427 commented:
asio has now been purged from this PR, and existing review comments have been addressed.  I'm going to take this PR out of draft.

FYI: on the 30th I will be leaving on vacation for a couple of weeks, so don't be concerned if I don't respond to review comments right away.

jmachowinski commented:
I'll try to have a look on it on Friday.

Yadunund commented:
@alsora / @jmachowinski friendly reminder to take another look at this PR. 🧇

jmachowinski commented on rclpy/src/rclpy/events_executor/delayed_event_thread.hpp:
```diff
@@ -0,0 +1,59 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
```
Sadly this is not allowed in the ros code base, as it's not part of the c++ standard

bmartin427 replied:
Got it, will switch to #ifndefs

bmartin427 replied:
Done.

jmachowinski commented on rclpy/src/rclpy/events_executor/events_executor.cpp:
```diff
@@ -0,0 +1,929 @@
+      waitables.attr("update")(py::set(node.attr("waitables")));
+
+      // It doesn't seem to be possible to support guard conditions with a callback-based (as
+      // opposed to waitset-based) API.  Fortunately we don't seem to need to.
```
guard conditions can be supported, and you need them for wakeup for the spin_until_future_complete (At least this the case for rclcpp).
 Normally you would just enqueue an empty event if one is triggered.

bmartin427 replied:
rcl doesn't seem to provide any way to know when a guard condition is triggered though, other than `rcl_wait()`?  There's potential to hack a callback into `rclpy.GuardCondition.trigger()`, though some of the guard conditions I might want to know about are triggered at the rcl level and not through rclpy so that doesn't help.

I also wasn't able to identify any particular need for guard conditions?  `spin_until_future_complete()` [works as-is](https://github.com/bmartin427/rclpy/blob/events_executor/rclpy/test/test_events_executor.py#L370-L374) by hooking the `Future` done callback.  Signal handling and wake-up upon new entities added are both done in different ways.  I'm not sure there's anything else?  Also as far as I can tell, rclcpp `EventsExecutor` [doesn't support them](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/experimental/executors/events_executor/events_executor.cpp#L399-L404) either.

alsora replied:
rclcpp guard conditions support the events executor, but it currently requires that they are triggered from the rclcpp layer and not from the rcl layer: https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/guard_condition.cpp

As you pointed out however, they are not hooked up in the executor.

The `spin_until_future_complete` relies on the executor base class implementation, where `spin_once_impl` is used and the future is handled separately

jmachowinski commented:
I looked a bit more over the code, but don't have the time right now to finish the review. Overall the code looks good and seems to be of high quality. 
The naming convention of the function names is off and might need changes.

Something I don't understand right now, is how the liefetime tracking of python objects works, and how the executer ensures that they are alive over the period of the call. @bmartin427 can you give some information on that ?
Actually, how do you destroy / remove a subscriber in the python world ?

jmachowinski commented:
@Flova you seem to work on a similar thing, would you mind taking a look at this ?

Flova commented:
Yeah this is a quite funny timing. I also did a WIP prototype of a python events executor, that greatly increased the performance. It currently is missing timers, a graceful shutdown, tests and proper formatting. It orients itself on the C++ one, but I also added support for callback groups in combination with spin until future complete, allowing to properly await e.g. a service result with a single thread. Fighting the GIL was quite challenging at time, leading to a lot of race conditions and dead locks, but it seems to be working now. It currently also supports stuff like adding subscriptions at runtime. You can find my current (not very clean) state here: https://github.com/ros2/rclpy/compare/iron...Flova:rclpy:feature/events_executor

I didn't have the time to look at this implementation here, but maybe we can combine both efforts in some way or another.

bmartin427 commented:
> The naming convention of the function names is off and might need changes.

You're talking about `snake_case` vs `CamelCase`?  I thought I understood that [either was acceptable](https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Code-Style-Language-Versions.html#function-and-method-naming), but I can switch if needed.

> Something I don't understand right now, is how the liefetime tracking of python objects works, and how the executer ensures that they are alive over the period of the call. @bmartin427 can you give some information on that ?

rclpy seems to have a convention where the python object owns the C++ `_rclpy` object, which supports context management; any sections during which the C++ object must not be destroyed must take place within a `with` section on the C++ object.  I created the C++ object [`ScopedWith`](https://github.com/bmartin427/rclpy/blob/events_executor/rclpy/src/rclpy/events_executor/scoped_with.hpp) so that I can do similar things like:
```
{
  ScopedWith with(object);
  // object can't be destroyed
}
// object may be destroyed
```
But then in practice, in most cases I found the need to protect the `_rclpy` object for as long as the executor is working with it (i.e. until `HandleRemoveFoo()`); so for that reason I keep the `ScopedWith` object around for longer than a single function scope by stashing it somewhere in the executor itself.

> Actually, how do you destroy / remove a subscriber in the python world ?

In Python you can't just let a subscription object go out of scope, you have to actually call [`Node.destroy_subscription()`](https://github.com/ros2/rclpy/blob/rolling/rclpy/rclpy/node.py#L1933), which removes the subscription from the node and sends a wakeup to the executor, which will notice the removal and clean up the callbacks and such.  Looking through test_events_executor.py now, although I test deleting a whole subscriber node, it looks like I don't have coverage for an existing node deleting one subscription, though I have no reason to think that won't work.  I can add that coverage for completeness' sake.

bmartin427 commented:
Sorry about the force-push, I'm trying to resolve this complaint from github:

> This branch must not contain merge commits.

This complaint persists after force-pushing a rebase, so I'm unsure what's wrong now.

Commit:
Swap 'pragma once' for ifndef include guards

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Add test coverage for Node.destroy_subscription()

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

bmartin427 commented:
> > Actually, how do you destroy / remove a subscriber in the python world ?
> 
> In Python you can't just let a subscription object go out of scope, you have to actually call [`Node.destroy_subscription()`](https://github.com/ros2/rclpy/blob/rolling/rclpy/rclpy/node.py#L1933), which removes the subscription from the node and sends a wakeup to the executor, which will notice the removal and clean up the callbacks and such. Looking through test_events_executor.py now, although I test deleting a whole subscriber node, it looks like I don't have coverage for an existing node deleting one subscription, though I have no reason to think that won't work. I can add that coverage for completeness' sake.

This coverage is now added.

bmartin427 commented:
I don't believe the test failures are my fault, and I'm unable to reproduce them locally.

@ros-pull-request-builder retest this please

alsora commented:
Overall this looks good!
I don't see any obvious problem...

As discussed in the client library WG, it may be nice to temporarily change the default executor to this events executor and try some complex python code such as the ROS CLI.
If this shows some problems, I think that they shouldn't necessarily be blockers to merge the PR, but at least we would be aware of the known bugs to fix.

alsora commented:
CI

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22663)](http://ci.ros2.org/job/ci_linux/22663/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16905)](http://ci.ros2.org/job/ci_linux-aarch64/16905/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=2179)](http://ci.ros2.org/job/ci_linux-rhel/2179/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23490)](http://ci.ros2.org/job/ci_windows/23490/)

alsora commented:
@bmartin427 unfortunately the CI errors are all related to the events-executor.
There are some unit-test failures in the RHEL CI (see https://ci.ros2.org/job/ci_linux-rhel/2179/testReport/) and compilation errors in windows (see https://ci.ros2.org/job/ci_windows/23490/console#console-section-13)

Commit:
Replace '|' type markup with typing.Optional and typing.Union

Python 3.9, still used by RHEL 9, doesn't seem to understand '|' syntax, and
Optional/Union seems to be what gets used throughout the rest of the codebase.

Additionally, fix a couple other mypy nits:
 * mypy is mad that I haven't explicitly annotated every __init__ as returning None
 * mypy wants generic arguments to service and action clients now

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

bmartin427 commented:
I think the RHEL issue is fixed.  I have yet to set up a Windows build environment to look into that.

bmartin427 commented:
I had gotten partway through the Windows setup process when evidently the process changed out from under me. :weary:
https://github.com/ros2/ros2_documentation/pull/4989

Commit:
Use 'auto' in place of explicit return type on hash

pybind11::hash() is documented as returning ssize_t, but this seems to be a lie because MSVC
doesn't understand that type; so, let's just return whatever we do get.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

Commit:
Change initialization of struct members

MSVC didn't like the more concise method.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

bmartin427 commented:
Ok, I believe all known CI issues are resolved now, though I'm not sure how to kick off a new round of tests.

alsora commented:
Only maintainers/committers can start a new CI.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22698)](http://ci.ros2.org/job/ci_linux/22698/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16937)](http://ci.ros2.org/job/ci_linux-aarch64/16937/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=2212)](http://ci.ros2.org/job/ci_linux-rhel/2212/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23546)](http://ci.ros2.org/job/ci_windows/23546/)

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-pmc-minutes-for-2025-03-12/42493/1

bmartin427 commented:
>     * Windows [![Build Status](https://camo.githubusercontent.com/1544ad09c8ee8e901715b30147f7b4cbc503e28617f47d19350341b6e20498c4/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f77696e646f7773266275696c643d3233353436)](http://ci.ros2.org/job/ci_windows/23546/)

I don't have any idea what that Windows failure is but I don't think it's my fault?

jmachowinski commented:
Yeah, this is just general Windows ci flakiness. I'll restart the windows build tomorrow, when I'm back at work.

jmachowinski commented:
https://ci.ros2.org/job/ci_windows/23585/

alsora commented:
Unfortunately there's still a test failure in windows for the events executor

https://ci.ros2.org/job/ci_windows/23585/testReport/junit/(root)/projectroot/test_executor/

```
            # Nothing is ever setting the future, so this should have waited
            # at least 0.1 seconds.
>           self.assertGreaterEqual(end - start, 0.1)
E           AssertionError: 0.0930000000007567 not greater than or equal to 0.1
```

bmartin427 commented:
> Unfortunately there's still a test failure in windows for the events executor
> 
> https://ci.ros2.org/job/ci_windows/23585/testReport/junit/(root)/projectroot/test_executor/
> 
> ```
>             # Nothing is ever setting the future, so this should have waited
>             # at least 0.1 seconds.
> >           self.assertGreaterEqual(end - start, 0.1)
> E           AssertionError: 0.0930000000007567 not greater than or equal to 0.1
> ```

Hmm. Was this existing test known to be flaky already? This wasn't seen in any of the previous runs nor in my own local testing. Also, checking for exactly the expected amount of realtime elapsed in a CI builder doesn't sound like a good idea anyway.

jmachowinski commented:
Hm, RunUntil does not enforce the timeout, might the queue run empty and therefore stops spinning to early ?

bmartin427 commented:
> Hm, RunUntil does not enforce the timeout, might the queue run empty and therefore stops spinning to early ?

Not sure what you mean, the deadline is passed to `std::condition_variable::wait_until()`, which should otherwise wait until the queue is stopped, or there's something to run.  This is done in a loop, so servicing something else wouldn't cause an early exit either.

https://github.com/bmartin427/rclpy/blob/events_executor/rclpy/src/rclpy/events_executor/events_queue.cpp#L40

I'll try to have a closer look at this tomorrow.

jmachowinski commented:
I reread the predicate and exit criteria about 5 times, and must agree with you, this should work. This failing test is a mystery to me, I don't see how this could happen. Memory corruption issue ?

Commit:
Use subTest in test_executor to distinguish which executor type failed

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

bmartin427 commented:
I've got this test running on a tight loop on my Windows machine, and I'm unable to reproduce any problem.  I've even tried setting it to low priority and thrashing it with other stuff in the foreground, with no issue on this test.  Furthermore, I can't even say for sure it was EventsExecutor and not SingleThreadedExecutor which failed this test, since both are run in the same test step.  To that end, I'm introducing a commit adding the use of `subTest` so that the executor type gets printed in the event of a failure.

bmartin427 committed:
Merge branch 'rolling' into events_executor

Signed-off-by: Brad Martin <52003535+bmartin427@users.noreply.github.com>
Signed-off-by: Brad Martin <bmartin@fatlxception.org>

alsora commented:
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23650)](http://ci.ros2.org/job/ci_windows/23650/)

bmartin427 commented:
Sigh, seems like it's still failing. Worse, it seems like something ate the subTest output, so I don't know any more than I did before. 😞

I'll think about this some more tomorrow.

bmartin427 commented:
I'm trying to run the actual Windows CI build scripts instead of just a regular build today, in case there are any relevant differences, however this is turning into a huge pain. I went through a bunch of Docker setup only to discover I can't run the docker on a Home version of Windows, I need Server. I may have things running now just skipping docker and directly invoking the scripts that would run in the container instead, but this problem may benefit from input from someone who knows more about Windows builds generally.

jmachowinski commented:
@nuclearsandwich do you have an good advice on the topic of how to run the CI on your 'normal' windows ?

jmachowinski commented:
@bmartin427 what about adding a second test, just running one executor as c&p, this would at least solve the question of which executor is the culprit.

bmartin427 commented:
I completed a run of the entire Windows CI (run_ros2_batch.py from ros2/ci), and got 9 errors, none of which were in rclpy:
```
==> env.bat C:\dev\.pixi\envs\default\Scripts\colcon.EXE test-result --test-result-base "build"
**********************************************************************
** Visual Studio 2019 Developer Command Prompt v16.11.44
** Copyright (c) 2021 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x86_x64'
build\demo_nodes_cpp\Testing\20250318-2048\Test.xml: 1 test, 0 errors, 1 failure, 0 skipped
build\demo_nodes_cpp\test_results\demo_nodes_cpp\test_tutorial_add_two_ints_server_add_two_ints_client_async__rmw_fastrtps_cpp.xunit.xml: 2 tests, 0 errors, 2 failures, 0 skipped
build\ros2bag\pytest.xml: 32 tests, 0 errors, 1 failure, 0 skipped
build\sros2\pytest.xml: 35 tests, 18 errors, 4 failures, 0 skipped
build\test_security\Testing\20250318-2120\Test.xml: 47 tests, 0 errors, 1 failure, 46 skipped

Summary: 45574 tests, 18 errors, 9 failures, 9539 skipped

II> colcon test-result returned: '1'
```
I also was still not entirely successful at reproducing the environment, because connextdds stuff denied me the permission to check it out.

> @bmartin427 what about adding a second test, just running one executor as c&p, this would at least solve the question of which executor is the culprit.

I can maybe look at reformulating the test yet again to get better debug output.  However, this has become a huge time sink for me with nothing to show for it so far.  I'm reaching the limit of the time I'm able to devote to trying to fix a problem in a build environment that's impossible to reproduce accurately.

bmartin427 commented:
Would it be possible for someone who is a Windows developer to try to reproduce this locally please?

Commit:
Use time.perf_counter() instead of time.monotonic() in executor test

time.monotonic() has a resolution of 16ms, which is way too coarse for the intervals
this test is trying to measure.

Signed-off-by: Brad Martin <bmartin@fatlxception.org>

bmartin427 commented:
I think I know the problem (although not necessarily why it only happens in CI).  The test is looking at the elapsed time as measured by `time.monotonic()`.  On Linux, `time.get_clock_info("monotonic").resolution` returns `1e9`, or nanosecond precision.  On Windows, the same expression returns `0.016`, or 16 *milliseconds*.  When this test fails, it does so by less than that amount.  I think the actual elapsed time is correct, but the read back time is just being rounded down to this large increment.

Internally, EventsExecutor is using `std::chrono::steady_clock`, which seems to have at least somewhat higher precision.  Invoking `std::chrono::duration_cast<std::chrono::duration<double>>(std::chrono::steady_clock::duration(1)).count()` returns 1e-9, also implying nanosecond resolution, although I'm not certain if that accurately describes the underlying clock or just the representation.

I have replaced all of the uses of `time.monotonic()` with `time.perf_counter()` in test_executor.py.  `get_clock_info()` says that clock is monotonic as well, so there shouldn't be any problems with time warping.  Interestingly, it seems `monotonic()` uses `GetTickCount64()` as its implementation, whereas `perf_counter()` uses `QueryPerformanceCounter()` as advertised; but on Python 3.13, it appears `monotonic()` has changed to use `QueryPerformanceCounter()` as well.

Anyway, @alsora could you kick off one more Windows build please?

fujitatomoya commented:
@bmartin427 

* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23661)](http://ci.ros2.org/job/ci_windows/23661/)

jmachowinski commented:
https://ci.ros2.org/job/ci_windows/23664/

jmachowinski commented:
rclpy.Clock with ClockType.STEADY_TIME seems like the more reasonable time source. This one is is known to return the correct time.

jmachowinski commented:
CI looks good, I don't think the rosbag test is related.

alsora commented:
Running a full CI again, just in case

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22832)](http://ci.ros2.org/job/ci_linux/22832/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=17038)](http://ci.ros2.org/job/ci_linux-aarch64/17038/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=2336)](http://ci.ros2.org/job/ci_linux-rhel/2336/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23692)](http://ci.ros2.org/job/ci_windows/23692/)

alsora commented:
Merging this, as discussed today in the client library WG.
The Windows failure in the rosbag test is common to also other `rclpy` PRs

Flova commented:
Any chance this will be back ported into jazzy?

bmartin427 commented:
I originally implemented it for Iron and then forward-ported it during the PR process, so there's no big reason why it couldn't work with Jazzy.  The biggest compatibility changes I ran into when porting was the addition of `TimerInfo` as an argument to timer callbacks, and a new `rcl` API for getting a timer's absolute expiration time instead of forcing me to deal with relative times.  I'm unsure if either of those things appeared in Jazzy or were new for Rolling.

That said, I myself am not planning to do any more porting work, but anyone else is welcome to. :slightly_smiling_face:

Flova commented:
Okay. I could try opening a backport PR, but would this be aligned with the backporting policy of new features (I am not familiar with it)? @alsora 

I am really looking forward to it. Performance was a major pain point when we migrated to ROS2 a few years ago, as many parts needed a cpp rewrite. Do you have any numbers on the performance gains of the new executor. What is the maximum throughput of this implementation? I also did a python port of the events executor concept  (you were faster and more polished in the end, so I discontinued it) and the maximum throughput of a node with 10 subscriptions increased from ~1khz to >10khz both running at 100% of one core iirc..

alsora commented:
@Flova that's great!
The main requirement for a backport is to not break API/ABI.
Looking at the changes here, that are almost entirely just adding new files, it should be doable!

bmartin427 commented:
> Do you have any numbers on the performance gains of the new executor. What is the maximum throughput of this implementation?

Have a look at #1389 , I posted a simple benchmark script and some results.

Flova commented:
Btw, thanks a lot for implementing this very much needed feature @bmartin427 !

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-pmc-minutes-for-march-25-2025/42807/1

mjcarroll commented:
@bmartin427 Just as one final follow-up, would you like to announce this feature to discourse?  I would also like to include this in the feature list for Kilted as many people will be interested in it.

n8hui commented:
Hi all, I've made an attempt to backport this to Humble (ros2/rclpy#1454) (ros2/rcl#1237). Please take a look if you have time!
